core/domain/auth_jobs_one_off.py

```diff
@@ -0,0 +1,188 @@
+# coding: utf-8
+#
+# Copyright 2020 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Auth-related one-off jobs."""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import ast
+import itertools
+
+from core import jobs
+from core.domain import auth_domain
+from core.platform import models
+from core.platform.auth import firebase_auth_services
+import python_utils
+
+import firebase_admin
+from firebase_admin import auth as firebase_auth
+from firebase_admin import exceptions as firebase_exceptions
+
+FIREBASE_IMPORT_USERS_MAX_LEN = 1000
+
+AUDIT_KEY = 'INFO: Pre-existing Firebase accounts'
+FAILURE_KEY = 'FAILURE: Failed to create Firebase accounts'
+SUCCESS_KEY = 'SUCCESS: Created Firebase accounts'
+WARNING_KEY = 'WARNING: No action needed'
+
+POPULATED_KEY = 'ALREADY_DONE'
+NOT_POPULATED_KEY = 'NEEDS_WORK'
+
+auth_models, user_models = (
+    models.Registry.import_models([models.NAMES.auth, models.NAMES.user]))
+
+
+class AuditUserEmailsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """One-off job to confirm whether every user has a unique email address."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [user_models.UserSettingsModel]
+
+    @staticmethod
+    def map(user):
+        yield (user.email, user.id)
+
+    @staticmethod
+    def reduce(email, user_ids):
+        if len(user_ids) > 1:
+            yield (
+                'ERROR: %s is a shared email' % email,
+                # NOTE: These are only sorted to make unit tests simpler.
+                ', '.join(sorted(user_ids)))
+
+
+class PopulateFirebaseAccountsOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """One-off job that maps over UserSettingsModels and imports them to Oppia's
+    Firebase server. The latter requires that we "specify [an ID] at a minimum",
+    so we use Oppia's user_id as the value. NOTE: THIS RELATIONSHIP **MUST NOT**
```

Comment:
I don't understand the note. Can you clarify it? What relationship is it talking about and why must it not be depended upon?